c/migrations: do not assert when building group partition map

mmaslankaprv opened the pull request:
When migrating consumer groups migration backend builds a map that associate consumer groups with `__consumer_offsets` topic partitions. The map is built when the migration is reconciled. Previously the race condition between application of state in migration and topic table could lead to a situation in which a `__consumer_offsets` partition is not yet present in partition manager. This is when assertion was triggered.

Added retry with small backoff to wait for the partition to appear instead of asserting and crashing Redpanda application.

Fixes: CORE-14421

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
c/migrations: do not assert when building group partition map

When migrating consumer groups migration backend builds a map that
associate consumer groups with `__consumer_offsets` topic partitions.
The map is built when the migration is reconciled. Previously the race
condition between application of state in migration and topic table
could lead to a situation in which a `__consumer_offsets` partition is
not yet present in partition manager. This is when assertion was
triggered.

Added retry with small backoff to wait for the partition to appear
instead of asserting and crashing Redpanda application.

Fixes: CORE-14421

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented:
I can see `build_migration_group_map` is called from `backend::get_entities_status` and from `backend::reconcile_migration`.

The call chain for the first one originates in Admin API, so it may hit a node unaware of consumer groups topic indeed. I would fix it with a call to `_group_proxy.assure_topic_exists`, similar to the one in `frontend::do_create_migration`. I guess the assert caught us here.

The second one is for a migration that has been successfully created and transferred to at least `preparing` state, and on the node that is aware of such migration and such transfer. Consumer group topic creation happens during migration creation. So the current node must be aware of consumer group topic metadata. I don't think we need to do anything here.

mmaslankaprv commented:
> I can see `build_migration_group_map` is called from `backend::get_entities_status` and from `backend::reconcile_migration`.
> 
> The call chain for the first one originates in Admin API, so it may hit a node unaware of consumer groups topic indeed. I would fix it with a call to `_group_proxy.assure_topic_exists`, similar to the one in `frontend::do_create_migration`. I guess the assert caught us here.
> 
> The second one is for a migration that has been successfully created and transferred to at least `preparing` state, and on the node that is aware of such migration and such transfer. Consumer group topic creation happens during migration creation. So the current node must be aware of consumer group topic metadata. I don't think we need to do anything here.

The assert was not from the admin api call but from the migration reconciliation: 
```
[Backtrace #6]
___interceptor_backtrace at /src/llvm-project/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:4497
seastar::guarded_backtrace(void**, int) at ./external/+non_module_dependencies+seastar/src/util/backtrace.cc:102
void seastar::backtrace<seastar::current_backtrace_tasklocal()::$_0>(seastar::current_backtrace_tasklocal()::$_0&&, bool) at ./external/+non_module_dependencies+seastar/include/seastar/util/backtrace.hh:101
 (inlined by) seastar::current_backtrace_tasklocal() at ./external/+non_module_dependencies+seastar/src/util/backtrace.cc:111
seastar::current_tasktrace() at ./external/+non_module_dependencies+seastar/src/util/backtrace.cc:148
seastar::current_backtrace() at ./external/+non_module_dependencies+seastar/src/util/backtrace.cc:181
detail::assert_failed_thunk2(char const*, char const*, fmt::v9::basic_format_args<fmt::v9::basic_format_context<fmt::v9::appender, char>>) at ./src/v/base/vassert.cc:103
void detail::assert_failed_thunk1<detail::base_named_type<seastar::basic_sstring<char, unsigned int, 15u, true>, kafka::kafka_group_id, std::__1::integral_constant<bool, false>> const&>(char const*, char const*, detail::base_named_type<seastar::basic_sstring<char, unsigned int, 15u, true>, kafka::kafka_group_id, std::__1::integral_constant<bool, false>> const&) at ./bazel-out/k8-dbg/bin/src/v/base/_virtual_includes/base/base/vassert.h:45
void detail::assert_failed_thunk0<detail::base_named_type<seastar::basic_sstring<char, unsigned int, 15u, true>, kafka::kafka_group_id, std::__1::integral_constant<bool, false>>>(char const*, char const*, detail::base_named_type<seastar::basic_sstring<char, unsigned int, 15u, true>, kafka::kafka_group_id, std::__1::integral_constant<bool, false>> const&) at ./bazel-out/k8-dbg/bin/src/v/base/_virtual_includes/base/base/vassert.h:55
 (inlined by) cluster::data_migrations::backend::build_migration_group_map(cluster::data_migrations::migration_metadata const&) const at ./src/v/cluster/data_migration_backend.cc:1789
cluster::data_migrations::backend::reconcile_migration(cluster::data_migrations::backend::migration_reconciliation_state&, cluster::data_migrations::migration_metadata const&) at ./src/v/cluster/data_migration_backend.cc:1806
cluster::data_migrations::backend::handle_migration_update(detail::base_named_type<long, cluster::data_migrations::data_migration_type_tag, std::__1::integral_constant<bool, true>>) (.resume) at ./src/v/cluster/data_migration_backend.cc:1386
std::__1::coroutine_handle<seastar::internal::coroutine_traits_base<void>::promise_type>::resume[abi:se200100]() const at ./external/toolchains_llvm++llvm+current_llvm_toolchain/bin/../../toolchains_llvm++llvm+current_llvm_toolchain_llvm/bin/../include/c++/v1/__coroutine/coroutine_handle.h:144
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ./external/+non_module_dependencies+seastar/include/seastar/core/coroutine.hh:122
seastar::reactor::task_queue::run_tasks() at ./external/+non_module_dependencies+seastar/src/core/reactor.cc:2721
seastar::reactor::task_queue_group::run_tasks() at ./external/+non_module_dependencies+seastar/src/core/reactor.cc:3222
seastar::reactor::task_queue_group::run_some_tasks() at ./external/+non_module_dependencies+seastar/src/core/reactor.cc:3206
seastar::reactor::do_run() at ./external/+non_module_dependencies+seastar/src/core/reactor.cc:3394
seastar::reactor::run() at ./external/+non_module_dependencies+seastar/src/core/reactor.cc:3267
seastar::app_template::run_deprecated(int, char**, std::__1::function<void ()>&&) at ./external/+non_module_dependencies+seastar/src/core/app-template.cc:273
seastar::app_template::run(int, char**, std::__1::function<seastar::future<int> ()>&&) at ./external/+non_module_dependencies+seastar/src/core/app-template.cc:167
application::run(int, char**) at ./src/v/redpanda/application.cc:540
main at ./src/v/redpanda/main.cc:22
/opt/redpanda_installs/ci/lib/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=cd410b710f0f094c6832edd95931006d883af48e, for GNU/Linux 3.2.0, stripped
```

mmaslankaprv committed:
c/controller_stm: wait for all topic metadata to applied from snapshot

When controller stm snapshot is applied the sub-stm snapshots are
applied in parallel, there are only a few of stms that state is applied
first f.e. features table. Moved topic table state to be applied before
other components. Other parts of Redpanda like migrations or cluster
linking may depend on the topic table to be up to date.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented on src/v/cluster/data_migration_backend.cc:
```diff
@@ -247,11 +247,26 @@ backend::get_entities_status(id migration_id) {
+              model::time_from_now(10s));
+            group_map_result = build_migration_group_map(meta);
+            --retry_cnt;
+        }
```
TBH I think we could only try once and return something like `errc::leadership_changed` to the API caller.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--07e416e7-b115-4539-97bd-af1e97d82570-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75536">build#75536</a></summary>


| test_class                    | test_method                          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                  |
|:------------------------------|:-------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_timestamps_match    | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "timestamp_type": "CreateTime"} | integration | https://buildkite.com/redpanda/redpanda/builds/75536#019a4ecb-456f-4529-ada9-73392d523c3f | FLAKY         | 20/21    | upstream reliability is '95.6896551724138'. current run reliability is '95.23809523809523'. drift is 0.45156 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| DataMigrationsApiTest         | test_creating_and_listing_migrations | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75536#019a4ecb-4573-4654-8f69-d4236ab393ec | FLAKY         | 17/21    | upstream reliability is '95.16757599376461'. current run reliability is '80.95238095238095'. drift is 14.2152 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations      |
| PartitionMovementTest         | test_empty                           | {"num_to_upgrade": 0}                                                                                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75536#019a4ecb-457c-455c-bae9-97dab43b2d97 | FLAKY         | 17/21    | upstream reliability is '85.10204081632654'. current run reliability is '80.95238095238095'. drift is 4.14966 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                                |
| PartitionMovementUpgradeTest  | test_basic_upgrade                   | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75536#019a4ecb-457e-4491-ac0f-2a2277c3cf00 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementUpgradeTest&test_method=test_basic_upgrade                 |
</details>
<!--/body-->
